File: 백준/Silver/2304. 창고 다각형/창고 다각형.py
import sys

input = sys.stdin.readline

n = int(input())

arr = [0 for i in range(1001)]

big_index = 0
last_index = 0

for i in range(n):
    a,b = map(int, input().split())
    arr[a] = b

    if arr[big_index] < b:
        big_index = a
    last_index = max(last_index, a)

ans = 0

now_val = arr[1]
for i in range(1, big_index):
    if arr[i] > now_val:
        ans += arr[i]
        now_val = arr[i]
    else:
        ans += now_val

now_val = arr[last_index]
for i in range(last_index, big_index, -1):
    if arr[i] > now_val:
        ans += arr[i]
        now_val = arr[i]
    else:
        ans += now_val

ans += arr[big_index]

print(ans)

File: 백준/Silver/2304. 창고 다각형/창고 다각형.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        int[] arr = new int[1001];

        int bigIndex = 0;
        int lastIndex = 0;
        int ans = 0;


        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int l = Integer.parseInt(st.nextToken());
            int h = Integer.parseInt(st.nextToken());

            arr[l] = h;

            if (arr[l] > arr[bigIndex]) {
                bigIndex = l;
            }
            lastIndex = Math.max(l, lastIndex);
        }

        int nowVal = arr[1];
        for (int i = 1; i < bigIndex; i++) {
            if (arr[i] > nowVal) {
                nowVal = arr[i];
            }
            ans += nowVal;
        }

        nowVal = arr[lastIndex];
        for (int i = lastIndex; i > bigIndex; i--) {
            if (arr[i] > nowVal) {
                nowVal = arr[i];
            }
            ans += nowVal;
        }
        ans += arr[bigIndex];
        System.out.println(ans);







    }
}